security/tasks.py

```diff
@@ -29,99 +31,257 @@ class LoggedTask(Task):
             stop=now(),
-            output=self.request.output_stream.getvalue()
+            output=self.request.output_stream.getvalue(),
+            retries=self.request.retries
+        )
+
+    def on_success(self, retval, task_id, args, kwargs):
+        self.on_success_task(self.task_log, args, kwargs, retval)
+
+    def on_failure_task(self, task_log, args, kwargs, exc):
+        """
+        On failure task is invoked if task end with some exception.
+        :param task_log: logged celery task instance
+        :param args: task args
+        :param kwargs: task kwargs
+        :param exc: raised exception which caused retry
+        """
+
+        LOGGER.log(self.logger_level, self.fail_error_message.format(
+            attempt=self.request.retries, exception=str(exc), task=task_log, task_name=task_log.name, **kwargs
+        ))
+        task_log.change_and_save(
+            state=CeleryTaskLogState.FAILED,
+            stop=now(),
+            error_message=str(exc),
+            output=self.request.output_stream.getvalue(),
+            retries=self.request.retries
+        )
+
+    def expire_task(self, task_log):
+        task_log.change_and_save(
+            state=CeleryTaskLogState.EXPIRED,
+            stop=now(),
+            error_message='Task execution was expired by command'
         )
-        self._call_callback('success', task_id, args, kwargs)
 
     def on_failure(self, exc, task_id, args, kwargs, einfo):
         try:
-            self.get_task().change_and_save(
-                state=CeleryTaskLogState.FAILED,
-                stop=now(),
-                error_message=einfo,
-                output=self.request.output_stream.getvalue()
-            )
+            self.on_failure_task(self.task_log, args, kwargs, exc)
         except CeleryTaskLog.DoesNotExist:
             pass
-        self._call_callback('failure', task_id, args, kwargs)
 
-    def _create_task(self, options, task_args, task_kwargs):
+    def _get_eta(self, options, now):
+        if options.get('countdown') is not None:
+            return now + timedelta(seconds=options['countdown'])
+        elif options.get('eta'):
+            return options['eta']
+        else:
+            return now
+
+    def _get_time_limit(self, options):
+        if options.get('time_limit') is not None:
+            return options['time_limit']
+        elif self.soft_time_limit is not None:
+            return self.soft_time_limit
+        else:
+            return self._get_app().conf.task_time_limit
+
+    def _get_stale_time_limit(self, stale_time_limit):
+        if stale_time_limit is not None:
+            return stale_time_limit
+        elif self.stale_time_limit is not None:
+            return self.stale_time_limit
+        elif hasattr(settings, 'CELERYD_TASK_STALE_TIME_LIMIT'):
+            return settings.CELERYD_TASK_STALE_TIME_LIMIT
+        else:
+            return None
+
+    def _get_expires(self, options, now, stale_time_limit):
```

Comment:
computation of expires is little bit complicated. If the value is not set we vill try to sompute from stale_time_limit and task time_limit

expires_time_limit = stale_time_limit - time_limit

it achieve that task cannot run if is set as stale